// react-native-reanimated/lib/module/index.js
'use strict';

import "./publicGlobals.js";
import { initializeReanimatedModule } from "./initializers.js";
import { ReanimatedModule } from './ReanimatedModule';

// TODO: Specify the initialization pipeline since now there's no
// universal source of truth for it.
// is-tree-shakable-suppress
initializeReanimatedModule(ReanimatedModule);

// eslint-disable-next-line import/first
import * as Animated from "./Animated.js";
export default Animated;
export { createAnimatedComponent } from "./Animated.js";
export { cancelAnimation, defineAnimation, GentleSpringConfig, GentleSpringConfigWithDuration, Reanimated3DefaultSpringConfig, Reanimated3DefaultSpringConfigWithDuration, SnappySpringConfig, SnappySpringConfigWithDuration, WigglySpringConfig, WigglySpringConfigWithDuration, withClamp, withDecay, withDelay, withRepeat, withSequence, withSpring, withTiming } from "./animation/index.js";
export { convertToRGBA, isColor } from "./Colors.js";
export { ReanimatedLogLevel } from "./common/index.js";
export { DynamicColorIOS, PlatformColor, processColor } from "./common/index.js";
export { InterfaceOrientation, IOSReferenceFrame, KeyboardState, ReduceMotion, SensorType } from "./commonTypes.js";
export { LayoutAnimationConfig } from "./component/LayoutAnimationConfig.js";
export { PerformanceMonitor } from "./component/PerformanceMonitor.js";
export { ReducedMotionConfig } from "./component/ReducedMotionConfig.js";
export { configureReanimatedLogger } from "./ConfigHelper.js";
export { enableLayoutAnimations, getViewProp, isConfigured, isReanimated3, makeMutable } from "./core.js";
export { NativeEventsManager } from "./createAnimatedComponent/index.js";
export * from "./css/index.js";
export { Easing } from "./Easing.js";
export { getDynamicFeatureFlag, getStaticFeatureFlag, setDynamicFeatureFlag } from "./featureFlags/index.js";
export { useAnimatedKeyboard, useAnimatedProps, useAnimatedReaction, useAnimatedRef, useAnimatedScrollHandler, useAnimatedSensor, useAnimatedStyle, useComposedEventHandler, useDerivedValue, useEvent, useFrameCallback, useHandler, useReducedMotion, useScrollOffset, /** @deprecated Please use {@link useScrollOffset} instead. */
useScrollOffset as useScrollViewOffset, useSharedValue, useTimestamp } from "./hook/index.js";
export { /** @deprecated Please use {@link Extrapolation} instead. */
Extrapolate, interpolateColor } from "./interpolateColor.js";
export { clamp, Extrapolation, interpolate } from "./interpolation.js";
export { isSharedValue } from "./isSharedValue.js";
export { advanceAnimationByFrame, advanceAnimationByTime, getAnimatedStyle, setUpTests, withReanimatedTimer } from './jestUtils';
export { BaseAnimationBuilder,
// Bounce
BounceIn, BounceInDown, BounceInLeft, BounceInRight, BounceInUp, BounceOut, BounceOutDown, BounceOutLeft, BounceOutRight, BounceOutUp, ComplexAnimationBuilder, CurvedTransition, EntryExitTransition,
// Fade
FadeIn, FadeInDown, FadeInLeft, FadeInRight, FadeInUp, FadeOut, FadeOutDown, FadeOutLeft, FadeOutRight, FadeOutUp, FadingTransition, FlipInEasyX, FlipInEasyY, FlipInXDown,
// Flip
FlipInXUp, FlipInYLeft, FlipInYRight, FlipOutEasyX, FlipOutEasyY, FlipOutXDown, FlipOutXUp, FlipOutYLeft, FlipOutYRight, JumpingTransition, Keyframe,
// Transitions
Layout, LightSpeedInLeft,
// Lightspeed
LightSpeedInRight, LightSpeedOutLeft, LightSpeedOutRight, LinearTransition,
// Pinwheel
PinwheelIn, PinwheelOut,
// Roll
RollInLeft, RollInRight, RollOutLeft, RollOutRight,
// Rotate
RotateInDownLeft, RotateInDownRight, RotateInUpLeft, RotateInUpRight, RotateOutDownLeft, RotateOutDownRight, RotateOutUpLeft, RotateOutUpRight, SequencedTransition, SharedTransition, SlideInDown, SlideInLeft,
// Slide
SlideInRight, SlideInUp, SlideOutDown, SlideOutLeft, SlideOutRight, SlideOutUp,
// Stretch
StretchInX, StretchInY, StretchOutX, StretchOutY,
// Zoom
ZoomIn, ZoomInDown, ZoomInEasyDown, ZoomInEasyUp, ZoomInLeft, ZoomInRight, ZoomInRotate, ZoomInUp, ZoomOut, ZoomOutDown, ZoomOutEasyDown, ZoomOutEasyUp, ZoomOutLeft, ZoomOutRight, ZoomOutRotate, ZoomOutUp } from "./layoutReanimation/index.js";
export { startMapper, stopMapper } from "./mappers.js";
export { jsVersion as reanimatedVersion } from "./platform-specific/jsVersion.js";
export { dispatchCommand, getRelativeCoords, measure, scrollTo, setGestureState, setNativeProps } from "./platformFunctions/index.js";
export { getUseOfValueInStyleWarning } from "./pluginUtils.js";
export { createAnimatedPropAdapter } from "./PropAdapters.js";
export { finishScreenTransition, ScreenTransition, startScreenTransition } from "./screenTransition/index.js";
export { createWorkletRuntime, executeOnUIRuntimeSync, isWorkletFunction, makeShareableCloneRecursive, runOnJS, runOnRuntime, runOnUI } from "./workletFunctions.js";
